js/compiler.ts

```diff
@@ -495,11 +509,71 @@ export class Compiler
       };
     });
   }
+}
 
-  private static _instance: Compiler | undefined;
+const compiler = new Compiler();
 
-  /** Returns the instance of `Compiler` or creates a new instance. */
-  static instance(): Compiler {
-    return Compiler._instance || (Compiler._instance = new Compiler());
+let startResMsg: msg.StartRes;
+
+/** Send to the privilaged side that we have setup and are ready. */
+function sendStart(): void {
+  const builder = flatbuffers.createBuilder();
+  msg.Start.startStart(builder);
+  const startOffset = msg.Start.endStart(builder);
+  const baseRes = sendSync(builder, msg.Any.Start, startOffset);
+  assert(baseRes != null);
+  assert(msg.Any.StartRes === baseRes!.innerType());
+  startResMsg = new msg.StartRes();
+  assert(baseRes!.inner(startResMsg) != null);
+}
+
+// set global objects for compiler web worker
+window.clearTimeout = clearTimer;
+window.console = console;
+window.postMessage = postMessage;
+window.setTimeout = setTimeout;
+window.workerMain = workerMain;
+window.close = workerClose;
+window.TextDecoder = TextDecoder;
+window.TextEncoder = TextEncoder;
+
+// provide the "main" function that will be called by the privilaged side when
+// lazy instantiating the compiler web worker
+window.compilerMain = function compilerMain() {
+  // workerMain should have already been called since a compiler is a worker.
+  const encoder = new TextEncoder();
+  const decoder = new TextDecoder();
+  compiler.recompile = startResMsg.recompileFlag();
+  log(`recompile ${compiler.recompile}`);
+  window.onmessage = ({ data }: { data: Uint8Array }) => {
+    const json = decoder.decode(data);
+    const lookup = JSON.parse(json) as CompilerLookup;
+
+    const moduleMetaData = compiler.compile(lookup.specifier, lookup.referrer);
+
+    const responseJson = JSON.stringify(moduleMetaData);
+    const response = encoder.encode(responseJson);
+    postMessage(response);
+  };
+};
+
+/* tslint:disable-next-line:no-default-export */
+export default function denoMain() {
+  libdeno.recv(handleAsyncMsgFromRust);
+
+  // First we send an empty `Start` message to let the privileged side know we
+  // are ready. The response should be a `StartRes` message containing the CLI
+  // args and other info.
+  sendStart();
+
+  setLogDebug(startResMsg.debugFlag());
+
+  // handle `--types`
+  // TODO(kitsonk) move to Rust fetching from compiler
+  if (startResMsg.typesFlag()) {
+    const defaultLibFileName = compiler.getDefaultLibFileName();
+    const defaultLibModule = compiler.resolveModule(defaultLibFileName, "");
+    console.log(defaultLibModule.sourceCode);
+    os.exit(0);
```

Comment:
We don't have to do this now, but I think it would be nice if somehow the type information could be passed back to Rust where it was ultimately printed.

Reply:
I am having a problem getting this to work at all right now, because I need to change it so that the compiler is spun up.  So I need to mess with Rust anyways, so will try to do that now.

Reply:
Solved tactically for now.